usermatch/backend.js

```diff
@@ -0,0 +1,166 @@
+    if (user) {
+      return res.status(200).json({ message: 'Login successful' });
+    } else {
+      return res.status(401).json({ error: 'Invalid credentials' });
+    }
+  } catch (error) {
+    return res.status(500).json({ error: error.message });
+  }
+});
+
+app.post('/userprofile', async (req, res) => {
+  const { username, password, name, age, interests, travel_spots, hobbies, working_out, latitude, longitude } = req.body;
+
+  try {
+    let user = await User.findOne({ username });
+    if (user) {
+      // Update existing user profile
+      user.name = name;
+      user.age = age;
+      user.interests = interests;
+      user.travel_spots = travel_spots;
+      user.hobbies = hobbies;
+      user.working_out = working_out;
+      user.location = {
+        type: 'Point',
+        coordinates: [parseFloat(longitude), parseFloat(latitude)],
+      };
+    } else {
+      // Create a new user profile
+      user = new User({ username, password, name, age, interests, travel_spots, hobbies, working_out, location: { type: 'Point', coordinates: [parseFloat(longitude), parseFloat(latitude)] } });
+    }
+
+    await user.save();
+    res.status(201).json({ message: 'Profile successfully created/updated' });
+  } catch (error) {
+    return res.status(500).json({ error: error.message });
+  }
+});
+
+app.post('/updatelocation', async (req, res) => {
+  const { username, latitude, longitude } = req.body;
+
+  try {
+    const user = await User.findOne({ username });
+    if (!user) {
+      return res.status(404).json({ error: 'User not found' });
+    }
+
+    // Update user's location
+    user.location = {
+      type: 'Point',
+      coordinates: [parseFloat(longitude), parseFloat(latitude)],
+    };
+
+    await user.save();
+    res.status(200).json({ message: 'Location updated successfully' });
+  } catch (error) {
+    return res.status(500).json({ error: error.message });
+  }
+});
+
+app.get('/interests', async (req, res) => {
+  const { username } = req.query;
+
+  try {
+    const user = await User.findOne({ username });
+    if (!user) {
+      return res.status(404).json({ error: 'User not found' });
+    }
+
+    // Find potential matches based on common interests
+    const potentialMatches = await User.find({
+      username: { $ne: username }, // Exclude the current user
+      interests: { $in: user.interests }, // Find users with at least one common interest
+    });
+
+    res.status(200).json(potentialMatches);
+  } catch (error) {
+    return res.status(500).json({ error: error.message });
+  }
+});
+
+app.get('/possiblematches', async (req, res) => {
+  const { username, latitude, longitude } = req.query;
+
+  try {
+    const currentUser = await User.findOne({ username });
+    if (!currentUser) {
+      return res.status(404).json({ error: 'User not found' });
+    }
+
+    // Find potential matches based on proximity, demographics, and user behavior
+    const potentialMatches = await User.find({
```

Comment:
Great job on this find filter! The format is very specific to mongoDB and is well-written here